### ISSUE TITLE
vmafossexec and ffmpeg with libvamf integrated got diffrent score

### ISSUE THREAD

**Issue by coolfat** [Netflix/vmaf#269]
I tried to compute vmaf score using ffmpeg integrated with libvmaf, and I compare the result with vmafossexec, but the score is different. 
I am sure that the vmaf version and the model file are the same. What is the reason?
The command is as follows:

- ffmpeg

 compute command:
./ffmpeg_g -pix_fmt yuv420p -s 1920x1080 -i demo_image.yuv -pix_fmt yuv420p -s 1920x1080 -i demo_image_jpg.yuv -lavfi libvmaf='log_path=./vmaf_ffmpeg_score.txt:psnr=1:ssim=1' -f null -

the result:
frame: 0, psnr: 28.261, ssim: 0.952, **adm: 0.882**, adm_num: 842.852, adm_den: 955.240, adm_num_scale0: 85.614, adm_den_scale0: 95.485, adm_num_scale1: 135.746, adm_den_scale1: 152.805, adm_num_scale2: 266.015, adm_den_scale2: 302.057, adm_num_scale3: 355.476, adm_den_scale3: 404.893, motion: 0.000, motion2: 0.000, vif_num_scale0: 4320724.500, vif_den_scale0: 9963059.000, vif_num_scale1: 1760851.250, vif_den_scale1: 2178718.750, vif_num_scale2: 465159.156, vif_den_scale2: 526156.750, vif_num_scale3: 119682.898, vif_den_scale3: 132543.703, vif: 0.521,
Generate final features (including derived atom features)...
Normalize features, SVM regression, denormalize score, clip...
frame: 0, adm2: 0.882346, adm_scale0: 0.896616, adm_scale1: 0.888362, adm_scale2: 0.880681, adm_scale3: 0.877952, motion: 0.000000, vif_scale0: 0.433674, vif_scale1: 0.808205, vif_scale2: 0.884070, vif_scale3: 0.902969, vif: 0.520794, motion2: 0.000000,
Exec FPS: 0.643893
VMAF score = 60.977099

- vmafossexec

 compute command:
./vmafossexec yuv420p 1920 1080 ../../ffmpeg-4.1/demo_image.yuv ../../ffmpeg-4.1/demo_image_jpg.yuv /usr/local/share/model/vmaf_v0.6.1.pkl --psnr --ssim

the result:
Start calculating VMAF score...
Read input model (pkl) at /usr/local/share/model/vmaf_v0.6.1.pkl ...
Read input model (libsvm) at /usr/local/share/model/vmaf_v0.6.1.pkl.model ...
Initialize storage arrays...
Extract atom features...
frame: 0, psnr: 28.261, ssim: 0.952, **adm: 1.004**, adm_num: 867.527, adm_den: 864.018, adm_num_scale0: 85.358, adm_den_scale0: 90.831, adm_num_scale1: 134.595, adm_den_scale1: 145.634, adm_num_scale2: 270.091, adm_den_scale2: 270.821, adm_num_scale3: 377.483, adm_den_scale3: 356.733, motion: 0.000, motion2: 0.000, vif_num_scale0: 4442769.000, vif_den_scale0: 9432217.000, vif_num_scale1: 1829720.875, vif_den_scale1: 2063461.375, vif_num_scale2: 492068.656, vif_den_scale2: 493145.156, vif_num_scale3: 127823.336, vif_den_scale3: 123463.047, vif: 0.569,
Generate final features (including derived atom features)...
Normalize features, SVM regression, denormalize score, clip...
frame: 0, adm2: 1.004061, adm_scale0: 0.939751, adm_scale1: 0.924202, adm_scale2: 0.997303, adm_scale3: 1.058168, motion: 0.000000, vif_scale0: 0.471021, vif_scale1: 0.886724, vif_scale2: 0.997817, vif_scale3: 1.035317, vif: 0.569041, motion2: 0.000000,
Exec FPS: 0.651152
VMAF score = 100.000000

**Comment by coolfat** [comment]
> This is unexpected, I'll take a look. Are you using the most recent versions of `libvmaf` and `ffmpeg`?

Yes, I am sure.

**Comment by fishjam** [comment]
@ChrisBeaumont @li-zhi 
I'm running into a similar issue, and I can reproduce this issue easily.

- Step:
  - 1.build debug and release version for vmafossexec in windows;
  - 2.run debug and release version with same parameters but with different out result

- Reason: debug's memory fill with 0xCD, release's memory is random.

![image](https://user-images.githubusercontent.com/1870684/51721174-8ecf3d00-208b-11e9-9a6d-e835769aaefb.png)

I find some suspicious code, please confirm it.
  1. [frame.c#L84](https://github.com/Netflix/vmaf/blob/master/feature/src/common/frame.c#L84) and [frame.c#L107](https://github.com/Netflix/vmaf/blob/master/feature/src/common/frame.c#L107), these code want read ref_rfile and dis_rfile to same temp_data. But I guess there should be offset in second fread, example: fread(temp_data+offset,  ...) ?

**Comment by li-zhi** [comment]
Here is the result produced on my Docker container, as comparison:
```
root@59defcacf663:/vmaf_oss/vmaf# ./wrapper/vmafossexec yuv420p 576 324 python/test/resource/yuv/src01_hrc00_576x324.yuv python/test/resource/yuv/src01_hrc01_576x324.yuv model/vmaf_v0.6.1.pkl --psnr --ssim --ms-ssim --thread 0
Start calculating VMAF score...
Exec FPS: 43.042209
VMAF score = 76.699271
PSNR score = 30.755064
SSIM score = 0.863227
MS-SSIM score = 0.963241
```
It looks like there is issue with the release version on Windows.

**Comment by HolyWu** [comment]
Both the release and debug version works fine here. Built from the latest git master branch by MSVC 2017 on Windows 10.

```
D:\Downloads\vmaf-master>x64\Debug\vmafossexec yuv420p 576 324 python\test\resource\yuv\src01_hrc00_576x324.yuv python\test\resource\yuv\src01_hrc01_576x324.yuv model\vmaf_v0.6.1.pkl --psnr --ssim --ms-ssim --thread 0
Start calculating VMAF score...
Exec FPS: 6.355521
VMAF score = 76.699271
PSNR score = 30.755064
SSIM score = 0.863227
MS-SSIM score = 0.963241

D:\Downloads\vmaf-master>x64\Release\vmafossexec yuv420p 576 324 python\test\resource\yuv\src01_hrc00_576x324.yuv python\test\resource\yuv\src01_hrc01_576x324.yuv model\vmaf_v0.6.1.pkl --psnr --ssim --ms-ssim --thread 0
Start calculating VMAF score...
Exec FPS: 21.810322
VMAF score = 76.699271
PSNR score = 30.755064
SSIM score = 0.863227
MS-SSIM score = 0.963241
```

**Comment by HolyWu** [comment]
I build again with v140 toolset (the toolset used by VS2015) instead of v141 toolset (the toolset used by VS2017 by default), and I can reproduce the wrong VMAF score now. After investigation, I find out the culprit is `convolution_avx.c` in project `libvmaf`. The project doesn't correctly set that file's instruction set to `/arch:AVX`, hence produces incorrect result. On the other hand, the project `vmafossexec` sets the entire project's instruction set to `/arch:AVX2`, which makes no sense at all.

**Comment by li-zhi** [comment]
@HolyWu does this mean the discrepancy is caused by whether AVX/AVX2 intrinsics is enabled/disabled?

**Comment by fishjam** [comment]
i don't think so. according coolfat’ description. He run the code in linux system or cgwin. It’s not same.

**Comment by HolyWu** [comment]
> 
> 
> @HolyWu does this mean the discrepancy is caused by whether AVX/AVX2 intrinsics is enabled/disabled?

The issue encountered by fishjam is indeed caused by AVX intrinsic is not enabled when compiled with Visual Studio 2015. But the cause of coolfat's problem is still unclear.

**Comment by Modeux** [comment]
I'm facing the same issue on mac.

My mac version is 10.14.2

My FFmpeg version is 4.1, installed via homebrew

My libvamf model version in FFmpeg : 

<img width="203" alt="2019-01-28 11 06 06" src="https://user-images.githubusercontent.com/22977327/51812807-216f1680-22ee-11e9-8d8f-2e358169f305.png">

My vmafossexec model version in vmaf github project: 

<img width="206" alt="2019-01-28 11 20 03" src="https://user-images.githubusercontent.com/22977327/51812958-b245f200-22ee-11e9-8f10-2d28b532c0ec.png">

Below is the description of how I used vmafossexec and libvmaf :

1. I have a 1080p test video, the video length is 30 seconds.

2. Using FFmpeg transcode the test video to 720p.

3. Upsample the 720p video to 1080p and convert to yuv420p format
`ffmpeg \
    -i test-30sec-720.mp4 \
    -vf scale=1920:1080 \
    -c:v rawvideo \
    -pix_fmt yuv420p \
    -sws_flags bicubic \
    test-30sec-1080.yuv`

4. Convert original 1080p video to yuv
`ffmpeg \
    -i test-orignal-30sec-1080.mp4 \
    -c:v rawvideo \
    -pix_fmt yuv420p \
    -sws_flags bicubic \
   test-orignal-30sec-1080.yuv`

5. Run vmafossexec
`wrapper/vmafossexec yuv420p 1920 1080 \
  test-orignal-30sec-1080.yuv \
  test-30sec-1080.yuv \
  model/vmaf_v0.6.1.pkl \
  --log vmaf_output.xml \
  --psnr --ssim --ms-ssim \
  --thread 0 --subsample 5`
  
   output:
       <img width="285" alt="2019-01-28 12 25 20" src="https://user-images.githubusercontent.com/22977327/51814764-d2c67a00-22f7-11e9-9294-24e92004c62a.png">

6. Run libvmaf
`ffmpeg -i test-30sec-720.mp4 \
       -i  test-orignal-30sec-1080.mp4 \
       -filter_complex "[0:v]scale=1920:1080[main];[main][1:v]libvmaf" \
       -f null -`
     output:
     <img width="447" alt="2019-01-28 12 29 52" src="https://user-images.githubusercontent.com/22977327/51814885-7dd73380-22f8-11e9-98ad-13ef997a8e96.png">

**Comment by christosbampis** [comment]
@Modeux can you share along the test file (mp4 and decoded) to try it out and debug this?

**Comment by HolyWu** [comment]
I compile FFmpeg with libvmaf today and do some tests. The problem only occurs when `scale` filter is used for upscaling the input before sending frame data to `libvmaf`. You can reproduce it easily even by the bundled test videos. Below are the steps. It looks like more a FFmpeg-side issue than a libvmaf-side issue.

1. Copy the two yuv videos to the root of vmaf folder so we needn't type long file path.
```
cp python/test/resource/yuv/src01_hrc00_576x324.yuv python/test/resource/yuv/src01_hrc01_576x324.yuv .
```

2. Upscale the two yuv videos to 1920x1080.
```
ffmpeg -f rawvideo -video_size 576x324 -i src01_hrc00_576x324.yuv -vf scale=1920:1080 -c:v rawvideo src01_hrc00_1920x1080.yuv
ffmpeg -f rawvideo -video_size 576x324 -i src01_hrc01_576x324.yuv -vf scale=1920:1080 -c:v rawvideo src01_hrc01_1920x1080.yuv
```

3. Calculate VMAF score by vmafossexec.
```
./wrapper/vmafossexec yuv420p 1920 1080 src01_hrc00_1920x1080.yuv src01_hrc01_1920x1080.yuv model/vmaf_v0.6.1.pkl

Start calculating VMAF score...
Exec FPS: 8.957663
VMAF score = 47.557533
```

4. Calculate VMAF score by FFmpeg without scale filter.
```
ffmpeg -f rawvideo -video_size 1920x1080 -i src01_hrc01_1920x1080.yuv -f rawvideo -video_size 1920x1080 -i src01_hrc00_1920x1080.yuv -lavfi libvmaf="model_path=model/vmaf_v0.6.1.pkl" -f null -

[libvmaf @ 0000025ac6a0e480] VMAF score: 47.557533
```

5. Calculate VMAF score by FFmpeg with scale filter.
```
ffmpeg -f rawvideo -video_size 576x324 -i src01_hrc01_576x324.yuv -f rawvideo -video_size 1920x1080 -i src01_hrc00_1920x1080.yuv -filter_complex "[0:v]scale=1920:1080[main];[main][1:v]libvmaf=model_path=model/vmaf_v0.6.1.pkl" -f null -

[libvmaf @ 0000022cf018a700] VMAF score: 36.482536
```

**Comment by li-zhi** [comment]
@slhck @kylophone any thoughts on this?

**Comment by kylophone** [comment]
My guess is the `swscale` scaling parameters are not the same when ffmpeg is invoked in this way. A quick way to check is run both ffmpeg commands with `-loglevel debug` and see what the `swscale` logs say.

**Comment by kylophone** [comment]
> My quick guess is the `swscale` scaling parameters are not the same when ffmpeg is invoked in this way. A quick way to check is run both ffmpeg commands with `-loglevel debug` and see what the `swscale` logs say.

Yes, a test on my setup shows me this is what is happening. There is a different default scaling algorithm for a normal filterchain (bicubic), and a complex filterchain (bilinear). To get what you're expecting you will need to explicitly set the scaling algorithm to something consistent.

**Comment by slhck** [comment]
> There is a different default scaling algorithm for a normal filterchain (bicubic), and a complex filterchain (bilinear)

That's very odd. I would have expected differences between using the `scale` filter and `-s` as an output option.

[Seems to be considered a bug](https://ffmpeg.org/pipermail/ffmpeg-user/2017-November/037911.html) but not yet fixed.

**Comment by HolyWu** [comment]
> @HolyWu, please verify this is the case for you as well. Try the same ffmpeg commands with `-loglevel debug` and see what swscale is doing.

Yes, it is. Adding `flags=bicubic` in `-filter_complex` produces the correct score now.

**Comment by Modeux** [comment]
Sorry for the late reply. Too busy on my working.

I test the libvmaf and update the argument.

`ffmpeg -loglevel debug \
       -i test-30sec-720.mp4 \
       -i test-original-30sec.mp4 \
       -filter_complex "[0:v]scale=1920x1080:flags=bicubic[main];[main][1:v]libvmaf" \
       -pix_fmt yuv420p \
       -f null -`

Output: VMAF score: 89.244752

And the vmafossexec VMAF score: 89.450385

The score is almost the same right now. But not sure why the score is still a little different.

Below is my test videos:
https://drive.google.com/drive/folders/1KbvoGOVMVgBLIxI9bXE-lT9rvQAdXrgL?usp=sharing

**Comment by HolyWu** [comment]
@Modeux No problem here. The score from your vmafossexec is wrong now. The score from FFmpeg is correct.

```
$ ffmpeg -i test-30sec-720.mp4 -vf scale=1920:1080 -c:v rawvideo test-30sec-1080.yuv
$ ffmpeg -i test-original-30sec.mp4 -c:v rawvideo test-original-30sec.yuv

$ ./wrapper/vmafossexec yuv420p 1920 1080 test-original-30sec.yuv test-30sec-1080.yuv model/vmaf_v0.6.1.pkl
Start calculating VMAF score...
Exec FPS: 9.248861
VMAF score = 89.244752

$ ffmpeg -f rawvideo -video_size 1920x1080 -i test-30sec-1080.yuv -f rawvideo -video_size 1920x1080 -i test-original-30sec.yuv -lavfi libvmaf="model_path=model/vmaf_v0.6.1.pkl" -f null -
[libvmaf @ 000001b4796fefc0] VMAF score: 89.244752

$ ffmpeg -i test-30sec-720.mp4 -i test-original-30sec.mp4 -filter_complex "[0:v]scale=1920:1080:flags=bicubic[main];[main][1:v]libvmaf=model_path=model/vmaf_v0.6.1.pkl" -f null -
[libvmaf @ 0000026fb0c5a840] VMAF score: 89.244752
```

**Comment by Modeux** [comment]
@HolyWu Thx, you are right.

I follow the step in vmafossexec.md. 

Copy past the command:

`wrapper/vmafossexec yuv420p 576 324 \
  python/test/resource/yuv/src01_hrc00_576x324.yuv \
  python/test/resource/yuv/src01_hrc01_576x324.yuv \
  model/vmaf_v0.6.1.pkl \
  --log vmaf_output.xml \
  --psnr --ssim --ms-ssim \
  --thread 0 --subsample 5`

the subsample argument cause the score more higher.

**Comment by slhck** [comment]
> I updated the FAQ and also added a section on how to use libvmaf with FFmpeg:
> https://github.com/Netflix/vmaf/blob/master/resource/doc/libvmaf.md#use-libvmaf-with-ffmpeg

I also updated the Scaling wiki page in FFmpeg to warn users about this. I have to say … that warning sign cannot be big enough for folks working in video quality research. I'll see if this bug can be patched.
